### PR TITLE
WIP: Generate the lists of resource based on the diff of two manifests

### PR DESCRIPTION
**Pull request by houshengbo**
Fixes #

## Proposed Changes

* This is the implementation of  a basic function, which returns the lists of resources to create, patch and delete, based on the delta between two manifests.

**Release Note**

cc @jcrossley3 @evankanderson @markusthoemmes

### CODE REVIEW

File: pkg/reconciler/common/manifest_diff_generator_test.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+import (
```
Comment (mattmoor) [inline pkg/reconciler/common/manifest_diff_generator_test.go:19]: Format Go code:
```suggestion
import (
	"testing"

```

File: pkg/reconciler/common/manifest_diff_generator_test.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+	mf "github.com/manifestival/manifestival"
+	util "knative.dev/operator/pkg/reconciler/common/testing"
+	"testing"
```
Comment (mattmoor) [inline pkg/reconciler/common/manifest_diff_generator_test.go:22]: Format Go code:
```suggestion

```

File: pkg/reconciler/common/manifest_diff_generator_test.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+	}
+}
+
+func notExistIn(t *testing.T, listResource []unstructured.Unstructured, manifest mf.Manifest) {
+	for _, val := range listResource {
+		found, resource := FindResourceByNSGroupKindName(val, manifest.Resources())
+		util.AssertEqual(t, found, false)
+		util.AssertDeepEqual(t, resource, unstructured.Unstructured{})
+	}
+}
```
Comment (mattmoor) [inline pkg/reconciler/common/manifest_diff_generator_test.go:64]: Add trailing newline:
```suggestion
}

```

File: pkg/reconciler/common/manifest_diff_generator.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+			listResourceDelete = append(listResourceDelete, resource)
+		}
+	}
+
+	return listResourceCreate, listResourcePatch, listResourceDelete
+}
+
+// FindResourceByNSGroupKindName is a lookup function, which returns true and the old resource, if the resource is available in an array of
+// unstructured.Unstructured, matching by namespace, APIVersion, Kind and Name.
+func FindResourceByNSGroupKindName(r unstructured.Unstructured, manifestResource []unstructured.Unstructured) (bool, unstructured.Unstructured)  {
```
Comment (mattmoor) [inline pkg/reconciler/common/manifest_diff_generator.go:56]: Format Go code:
```suggestion
func FindResourceByNSGroupKindName(r unstructured.Unstructured, manifestResource []unstructured.Unstructured) (bool, unstructured.Unstructured) {
```

File: pkg/reconciler/common/manifest_diff_generator_test.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+import (
```
Comment (mattmoor) [inline pkg/reconciler/common/manifest_diff_generator_test.go:19]: Format Go code:
```suggestion
import (
	"testing"

```

File: pkg/reconciler/common/manifest_diff_generator_test.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"testing"
```
Comment (mattmoor) [inline pkg/reconciler/common/manifest_diff_generator_test.go:21]: Format Go code:
```suggestion

```

File: pkg/reconciler/common/manifest_diff_generator_test.go [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+		util.AssertDeepEqual(t, val, resource)
+	}
+}
+
+func notExistIn(t *testing.T, listResource []unstructured.Unstructured, manifest mf.Manifest) {
+	for _, val := range listResource {
+		found, resource := FindResourceByNSGroupKindName(val, manifest.Resources())
+		util.AssertEqual(t, found, false)
+		util.AssertDeepEqual(t, resource, unstructured.Unstructured{})
+	}
```
Comment (mattmoor) [inline pkg/reconciler/common/manifest_diff_generator_test.go:63]: Format Go code:
```suggestion
	}
}
```

File: pkg/reconciler/common/manifest_diff_generator.go (hunk 4 above)
Comment (mattmoor) [inline pkg/reconciler/common/manifest_diff_generator.go:56]: Format Go code:
```suggestion
func FindResourceByNSGroupKindName(r unstructured.Unstructured, manifestResource []unstructured.Unstructured) (bool, unstructured.Unstructured) {
```

File: pkg/reconciler/common/manifest_diff_generator_test.go [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+import (
+	"testing"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
```
Comment (mattmoor) [inline pkg/reconciler/common/manifest_diff_generator_test.go:22]: Format Go code:
```suggestion

```

File: pkg/reconciler/common/manifest_diff_generator_test.go [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+	mf "github.com/manifestival/manifestival"
```
Comment (mattmoor) [inline pkg/reconciler/common/manifest_diff_generator_test.go:23]: Format Go code:
```suggestion
	mf "github.com/manifestival/manifestival"
	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
```

File: pkg/reconciler/common/manifest_diff_generator_test.go (hunk 8 above)
Comment (mattmoor) [inline pkg/reconciler/common/manifest_diff_generator_test.go:22]: Format Go code:
```suggestion

```

File: pkg/reconciler/common/manifest_diff_generator_test.go (hunk 9 above)
Comment (mattmoor) [inline pkg/reconciler/common/manifest_diff_generator_test.go:23]: Format Go code:
```suggestion
	mf "github.com/manifestival/manifestival"
	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
```

File: pkg/reconciler/common/manifest_diff_generator_test.go [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+*/
+
+package common
+
+import (
+	"testing"
+
+	mf "github.com/manifestival/manifestival"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
```
Comment (mattmoor) [inline pkg/reconciler/common/manifest_diff_generator_test.go:24]: Format Go code:
```suggestion

```

Comment (mattmoor) [inline pkg/reconciler/common/manifest_diff_generator_test.go:24]: Format Go code:
```suggestion

```

**Comment by jcrossley3** [comment]
/hold

I have some reservations about this approach. It seems overly complex to me. Let's discuss after the holidays.

File: pkg/reconciler/common/manifest_diff_generator.go [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+import (
+	"context"
+	"fmt"
+	"k8s.io/client-go/rest"
```
Comment (mattmoor) [inline pkg/reconciler/common/manifest_diff_generator.go:22]: Format Go code:
```suggestion

```

File: pkg/reconciler/common/manifest_diff_generator.go [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+	"os"
+	"path/filepath"
```
Comment (mattmoor) [inline pkg/reconciler/common/manifest_diff_generator.go:24]: Format Go code:
```suggestion
	"path/filepath"

	"k8s.io/client-go/rest"
```

File: pkg/reconciler/common/manifest_diff_generator.go [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+
+	"github.com/go-logr/zapr"
+	mfc "github.com/manifestival/client-go-client"
+	mf "github.com/manifestival/manifestival"
+	"go.uber.org/zap"
+	"knative.dev/pkg/logging"
```
Comment (mattmoor) [inline pkg/reconciler/common/manifest_diff_generator.go:30]: Format Go code:
```suggestion

```

File: pkg/reconciler/common/manifest_diff_generator.go [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
```
Comment (mattmoor) [inline pkg/reconciler/common/manifest_diff_generator.go:31]: Format Go code:
```suggestion
	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
	"knative.dev/pkg/logging"
```

File: pkg/reconciler/knativeserving/controller.go [hunk 15]
```diff
@@ -16,19 +16,13 @@ package knativeserving
 import (
 	"context"
 	"fmt"
```
Comment (mattmoor) [inline pkg/reconciler/knativeserving/controller.go:18]: Format Go code:
```suggestion
	"fmt"

```

File: pkg/reconciler/knativeserving/controller.go [hunk 16]
```diff
@@ -64,19 +58,11 @@ func NewController(ctx context.Context, cmw configmap.Watcher) *controller.Impl
...
-	if err != nil {
-		logger.Fatalw("Error creating the Manifest for knative-serving", zap.Error(err))
-	}
-
 	c := &Reconciler{
 		kubeClientSet:     kubeClient,
 		operatorClientSet: operatorclient.Get(ctx),
 		platform:          common.GetPlatforms(ctx),
-		config:            config,
+		config:		       injection.GetConfig(ctx),
```
Comment (mattmoor) [inline pkg/reconciler/knativeserving/controller.go:65]: Format Go code:
```suggestion
		config:            injection.GetConfig(ctx),
```

File: pkg/reconciler/knativeserving/knativeserving.go [hunk 17]
```diff
@@ -19,8 +19,8 @@ package knativeserving
 import (
 	"context"
 	"fmt"
```
Comment (mattmoor) [inline pkg/reconciler/knativeserving/knativeserving.go:21]: Format Go code:
```suggestion
	"fmt"

```

File: pkg/reconciler/knativeserving/knativeserving.go [hunk 18]
```diff
@@ -48,9 +48,14 @@ type Reconciler struct {
...
 	// config is the manifest of KnativeServing
-	config mf.Manifest
+	config *rest.Config
+	// oldManifest is the previous manifest of KnativeServing
+	oldManifest mf.Manifest
+	// currentManifest is the current manifest of KnativeServing
+	currentManifest mf.Manifest
 	// Platform-specific behavior to affect the transform
 	platform common.Platforms
+
```
Comment (mattmoor) [inline pkg/reconciler/knativeserving/knativeserving.go:58]: Format Go code:
```suggestion

```

**Pull request closed without merging**